File: app/src/main/java/com/angogasapps/myfamily/async/notification/FcmMessageBuilder.java
package com.angogasapps.myfamily.async.notification;

import android.content.Context;
import android.util.Log;


import com.angogasapps.myfamily.R;
import com.angogasapps.myfamily.app.AppApplication;
import com.angogasapps.myfamily.objects.Message;
import com.angogasapps.myfamily.utils.StringFormater;

import org.json.JSONObject;

import java.util.HashMap;

import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.TYPE_IMAGE_MESSAGE;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.TYPE_TEXT_MESSAGE;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.TYPE_VOICE_MESSAGE;

public class FcmMessageBuilder {

    private static final String NODE_DATA = "data";
    private static final String NODE_NOTIFICATION = "notification";
    private static final String CHILD_TOPIC = "topic";
    private static final String CHILD_TITLE = "title";
    private static final String CHILD_BODY = "body";

    private static final String CHILD_IMAGE = "image";


    private String to = "";
    private String title = "";
    private String body = "";
    private String image = "";
    private HashMap<String, String> data = new HashMap<>();

    public FcmMessageBuilder setTo(String to) {
        this.to = to;
        return this;
    }

    public FcmMessageBuilder setTitle(String title) {
        this.title = title;
        return this;
    }

    public FcmMessageBuilder setBody(String body) {
        this.body = body;
        return this;
    }

    public FcmMessageBuilder setData(HashMap<String, String> data) {
        this.data = data;
        return this;
    }

    public FcmMessageBuilder setImage(String imageUrl) {
        this.image = imageUrl;
        return this;
    }


    public JSONObject build() {
        return buildMessage();
    }

    private JSONObject buildMessage() {
        try {
            JSONObject mainObj = new JSONObject();
            JSONObject messageObj = new JSONObject();
            JSONObject notificationObj = new JSONObject();
            JSONObject dataObj = new JSONObject();

            notificationObj.put(CHILD_TITLE, StringFormater.convertStringToUTF8(this.title));
            notificationObj.put(CHILD_BODY, StringFormater.convertStringToUTF8(this.body));
            notificationObj.put(CHILD_IMAGE, this.image);

            for (String key : this.data.keySet())
                dataObj.put(key, this.data.get(key));

            messageObj.put(CHILD_TOPIC, this.to);
            messageObj.put(NODE_NOTIFICATION, notificationObj);
            messageObj.put(NODE_DATA, dataObj);

            mainObj.put("message", messageObj);
            return mainObj;
        } catch (Exception e) {
            Log.e("tag", "ERROR on build FcmMessage");
            return null;
        }
    }
}

File: app/src/main/java/com/angogasapps/myfamily/async/notification/FcmMessageBuilder.kt
package com.angogasapps.myfamily.async.notification

import android.util.Log
import com.angogasapps.myfamily.utils.StringFormater.convertStringToUTF8
import com.angogasapps.myfamily.async.notification.FcmMessageBuilder
import org.json.JSONObject
import com.angogasapps.myfamily.utils.StringFormater
import com.google.gson.JsonObject
import java.lang.Exception
import java.util.HashMap

class FcmMessageBuilder {
    private var to = ""
    private var title = ""
    private var body = ""
    private var image = ""
    private var data = HashMap<String, String>()
    fun setTo(to: String): FcmMessageBuilder {
        this.to = to
        return this
    }

    fun setTitle(title: String): FcmMessageBuilder {
        this.title = title
        return this
    }

    fun setBody(body: String): FcmMessageBuilder {
        this.body = body
        return this
    }

    fun setData(data: HashMap<String, String>): FcmMessageBuilder {
        this.data = data
        return this
    }

    fun setImage(imageUrl: String): FcmMessageBuilder {
        image = imageUrl
        return this
    }

    fun build(): JSONObject {
        return buildMessage()
    }

    private fun buildMessage(): JSONObject {
        return try {
            val mainObj = JSONObject()
            val messageObj = JSONObject()
            val notificationObj = JSONObject()
            val dataObj = JSONObject()
            notificationObj.put(
                CHILD_TITLE, convertStringToUTF8(
                    title
                )
            )
            notificationObj.put(
                CHILD_BODY, convertStringToUTF8(
                    body
                )
            )
            notificationObj.put(CHILD_IMAGE, image)
            for (key in data.keys) dataObj.put(key, data[key])
            messageObj.put(CHILD_TOPIC, to)
            messageObj.put(NODE_NOTIFICATION, notificationObj)
            messageObj.put(NODE_DATA, dataObj)
            mainObj.put(CHILD_MESSAGE, messageObj)
            mainObj
        } catch (e: Exception) {
            Log.e("tag", "ERROR on build FcmMessage")
            JSONObject()
        }
    }

    companion object {
        private const val NODE_DATA = "data"
        private const val NODE_NOTIFICATION = "notification"
        const val CHILD_TOPIC = "topic"
        private const val CHILD_TITLE = "title"
        private const val CHILD_BODY = "body"
        const val CHILD_MESSAGE = "message"
        private const val CHILD_IMAGE = "image"
    }
}